src/Core/AwsClient.php

```diff
@@ -4,33 +4,93 @@
 
 namespace AsyncAws\Core;
 
+use AsyncAws\Core\Credentials\CacheProvider;
+use AsyncAws\Core\Credentials\ChainProvider;
+use AsyncAws\Core\Credentials\ConfigurationProvider;
+use AsyncAws\Core\Credentials\CredentialProvider;
+use AsyncAws\Core\Credentials\IniFileProvider;
+use AsyncAws\Core\Credentials\InstanceProvider;
+use AsyncAws\Core\Exception\InvalidArgument;
 use AsyncAws\Core\Exception\MissingDependency;
-use AsyncAws\Core\Exception\RuntimeException;
 use AsyncAws\Core\Sts\StsClient;
+use AsyncAws\Lambda\LambdaClient;
 use AsyncAws\S3\S3Client;
 use AsyncAws\Ses\SesClient;
+use AsyncAws\Sns\SnsClient;
 use AsyncAws\Sqs\SqsClient;
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+use Symfony\Component\HttpClient\HttpClient;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
 
 /**
  * Base API client that instantiate other API classes if needed.
  *
  * @author Tobias Nyholm <tobias.nyholm@gmail.com>
  */
-class AwsClient extends AbstractApi
+class AwsClient
 {
     /**
      * @var array
      */
     private $serviceCache;
 
-    public function sts(): StsClient
+    /**
+     * @var HttpClientInterface
+     */
+    private $httpClient;
+
+    /**
+     * @var Configuration
+     */
+    private $configuration;
+
+    /**
+     * @var CredentialProvider
+     */
+    private $credentialProvider;
+
+    /**
+     * @var LoggerInterface|null
+     */
+    private $logger;
+
+    /**
+     * @param Configuration|array $configuration
+     */
+    public function __construct($configuration = [], ?CredentialProvider $credentialProvider = null, ?HttpClientInterface $httpClient = null, ?LoggerInterface $logger = null)
     {
-        if (!class_exists(StsClient::class)) {
-            throw MissingDependency::create('async-aws/core', 'STS');
+        if (\is_array($configuration)) {
+            $configuration = Configuration::create($configuration);
+        } elseif (!$configuration instanceof Configuration) {
+            throw new InvalidArgument(sprintf('Second argument to "%s::__construct()" must be an array or an instance of "%s"', __CLASS__, Configuration::class));
+        }
+
+        $this->httpClient = $httpClient ?? HttpClient::create();
+        $this->logger = $logger ?? new NullLogger();
+        $this->configuration = $configuration;
+        if ($configuration->isDefault(Configuration::OPTION_ENDPOINT)) {
+            $this->credentialProvider = $credentialProvider ?? new CacheProvider(new ChainProvider([
+                new ConfigurationProvider(),
+                new IniFileProvider($this->logger),
+                new InstanceProvider($this->httpClient, $this->logger),
+            ]));
+        } else {
+            $this->credentialProvider = $credentialProvider ?? new CacheProvider(new ChainProvider([
+                new ConfigurationProvider(),
+                new IniFileProvider($this->logger),
```

Comment:
Could we refactor this constructor code somehow?

Reply:
fixed by #190

Reply:
remove the `else` part and just keep the `if`

Reply:
Like this?

Reply:
:+1: